### PR TITLE
Fix offchain workers and add tests

### LINKED ISSUE #1405: Unit tests for offchain workers
Unsafe CDPs should be liquidated by cdp-engine offchain worker

### PR DESCRIPTION
**Pull request by ferrell-code**
Closes #1405 

Fixes auction-manager and cdp-engine offchain workers and adds tests

### CODE REVIEW

File: modules/cdp-engine/src/lib.rs [hunk 1]
```diff
@@ -648,23 +648,29 @@ impl<T: Config> Pallet<T> {
...
 		let is_shutdown = T::EmergencyShutdown::is_shutdown();
-		let mut map_iterator =
-			<loans::Positions<T>>::iter_prefix_from(currency_id, start_key.clone().ok_or(OffchainErr::OffchainStore)?);
+
+		// If start key is Some(value) continue iterating from that point in storage otherwise start
+		// iterating from the beginning of <loans::Positons<T>>
+		let mut map_iterator = match start_key.clone() {
+			Some(key) => <loans::Positions<T>>::iter_prefix_from(currency_id, key),
+			None => <loans::Positions<T>>::iter_prefix(currency_id),
+		};
```
Comment (ferrell-code) [inline modules/cdp-engine/src/lib.rs:661]: Mistake 1: when start key is none that is not an error it just means start iterating from the beginning

File: modules/cdp-engine/src/lib.rs [hunk 2]
```diff
@@ -648,23 +648,29 @@ impl<T: Config> Pallet<T> {
 		// get the max iterationns config
 		let max_iterations = StorageValueRef::persistent(OFFCHAIN_WORKER_MAX_ITERATIONS)
 			.get::<u32>()
-			.unwrap_or(Some(DEFAULT_MAX_ITERATIONS))
-			.ok_or(OffchainErr::OffchainStore)?;
+			.unwrap_or(Some(DEFAULT_MAX_ITERATIONS));
```
Comment (ferrell-code) [inline modules/cdp-engine/src/lib.rs:651]: Mistake 2: When max iterations is stored as None this is not an error, just means iterator will run to completion no matter the size of the iterator. This is probably not what collators would want to set but is still a valid state

File: modules/cdp-engine/src/lib.rs [hunk 3]
```diff
@@ -708,7 +714,7 @@ impl<T: Config> Pallet<T> {
 				};
 			to_be_continue.set(&(next_collateral_position, Option::<Vec<u8>>::None));
 		} else {
-			to_be_continue.set(&(collateral_position, Some(map_iterator.last_raw_key())));
+			to_be_continue.set(&(collateral_position, Some(map_iterator.prefix())));
```
Comment (ferrell-code) [inline modules/cdp-engine/src/lib.rs:717]: Mistake 3: We should store the current raw key, Not the previous one

**Comment by ferrell-code** [comment]
Now there are unit tests so no one else can break it :)

File: modules/cdp-engine/src/lib.rs (hunk 2 above)
Comment (xlc) [reply modules/cdp-engine/src/lib.rs:651]: I think we want to make None to be `DEFAULT_MAX_ITERATIONS`

Comment (ferrell-code) [reply modules/cdp-engine/src/lib.rs:651]: Yeah that makes sense, It is kinda a foot-gun unnecessarily for node operators

File: modules/auction-manager/src/lib.rs [hunk 4]
```diff
@@ -351,7 +351,8 @@ impl<T: Config> Pallet<T> {
 		// get the max iterationns config
 		let max_iterations = StorageValueRef::persistent(OFFCHAIN_WORKER_MAX_ITERATIONS)
 			.get::<u32>()
-			.unwrap_or(Some(DEFAULT_MAX_ITERATIONS));
+			.unwrap_or(Some(DEFAULT_MAX_ITERATIONS))
+			.unwrap_or(DEFAULT_MAX_ITERATIONS);
```
Comment (xlc) [inline modules/auction-manager/src/lib.rs:355]: should have some tests for this change

**Review by xlc — DISMISSED** [review]

**Comment by ferrell-code** [comment]
Don't merge this yet, key still doesn't work to continue iterator from where it left off, I will make add to test so it fails in that case, when I figure out how to fix it

File: modules/cdp-engine/src/tests.rs [hunk 5]
```diff
@@ -922,3 +941,192 @@ fn close_cdp_has_debit_by_swap_on_alternative_path() {
...
+			Change::NoChange,
+			Change::NoChange,
+			Change::NoChange,
+		));
+		run_to_block_offchain(2);
+		// should only run 1000 iterations stopping due to DEFAULT_MAX_ITERATIONS
+		assert_eq!(pool_state.write().transactions.len(), 1000);
+		// should only now run 1 iteration to finish off where it ended last block
+		run_to_block_offchain(3);
+		assert_eq!(pool_state.write().transactions.len(), 1001);
```
Comment (ferrell-code) [inline modules/cdp-engine/src/tests.rs:1130]: this tests that iterator correctly takes off where it was stopped, the only way for it to get 1001 tx is if the iterator is set to the 1000 on the second offchain worker

**Review by xlc — APPROVED** [review]

**Pull request merged**